### ISSUE TITLE
nan fillvalue attributes written by xarray

### ISSUE THREAD

**Issue by veenstrajelmer** [Deltares/xugrid#176]
I noticed that xugrid writes nan fillvalue attributes for variables without fillvalue defined. After a quick assessment I see that this is xarray behaviour and it is also happening in their code. A nan fillvalue does not make sense to me and it causes issues in some cases, like [this one](https://github.com/Deltares/dfm_tools/issues/653). In dfm_tools there is a workaround in place that pops nan fillvalue attrs upon reading, but I believe we should avoid writing them. It might be related to xarray not decoding default fillvalues if no fillvalue attribute is present. Either way, let's discuss what is expected/desired behaviour here.

The code below reads a D-FlowFM output which has no nan fillvalue attributes. When writing it with `.to_netcdf()` there are nan fillvalue attributes. I have tried this also with xugrid testdata, but those datasets already contain nan fillvalue attributes upon reading:
```python
import dfm_tools as dfmt
import xarray as xr
import numpy as np

file_nc = dfmt.data.fm_curvedbend_map(return_filepath=True)
file_out = "temp_fillvals_map.nc"
ds_org = xr.open_dataset(file_nc)
ds_org.to_netcdf(file_out)
ds_out_xr = xr.open_dataset(file_out)

print("nan fillvalue attrs in original dataset:")
ds = ds_org
count_xr = 0
for varn in ds.variables:
    if '_FillValue' in ds[varn].encoding:
        if np.isnan(ds[varn].encoding['_FillValue']):
            print(varn, ds[varn].encoding['_FillValue'])
            count_xr += 1

print()

print("nan fillvalue attrs in dataset written by xugrid/xarray:")
ds = ds_out_xr
count_xr = 0
for varn in ds.variables:
    if '_FillValue' in ds[varn].encoding:
        if np.isnan(ds[varn].encoding['_FillValue']):
            print(varn, ds[varn].encoding['_FillValue'])
            count_xr += 1
```

**Comment by JoerivanEngelen** [comment]
Just checking @veenstrajelmer, does https://github.com/pydata/xarray/pull/8713 fix this issue?

**Comment by veenstrajelmer** [comment]
I do not think so, this is about (lack of) precision of scale_factor attrs and is unrelated to {"_FillValue":np.nan} in the netcdf file attributes. I found the list of issues @Huite and myself collected a while ago. They are listed in this issue: https://github.com/Deltares/dfm_tools/issues/490

**Comment by kmuehlbauer** [comment]
@JoerivanEngelen @veenstrajelmer Just stepping by from https://github.com/pydata/xarray/pull/8713. Correct, that PR won't fix this particular issue. But I'm aiming for solving the default fillvalue issue over at xarray, too. This is moving in small steps, though.

**Comment by veenstrajelmer** [comment]
@kmuehlbauer thank you for clarifying this. We were considering to fix it in this package (prevent writing of `_Fillvalue:np.nan` attributes and add encoding of default fillvalues per datatype if no _FillValue is supplied, both in line with the cf-convention. Do you mean that this behaviour will in the future be taken care of by xarray? In that case we can indeed consider to temporarily put it in xugrid to anticipate future behaviour.

**Comment by Huite** [comment]
I'm going to close this issue. The NaN fill value is an xarray affair, and therefore doesn't really relate to UGRID.

You can argue that NaN isn't ideal since you cannot just check `data == FillValue`, but CF conventions don't disallow it, see: https://cfconventions.org/Data/Trac-tickets/52.html

* NaN is float
* CF-conventions do not forbid the usage of NaN as FillValue
* Everything which is not forbidden is allowed
* FillValue = NaN is allowed
* Any simulation code should support it, or prohibit it itself --- in which case it isn't CF-compliant!
    - In which case tools between xarray and the simulation code (like dfm_tools) will need to cover the discrepancy

That the encoding isn't properly updated (#321) does seem like a xugrid issue...

**Comment by veenstrajelmer** [comment]
I agree, this is not a xugrid issue, but a xarray issue. It is inconvenient, since the original files use [default fillvalues](https://github.com/Unidata/netcdf4-python/blob/9426a684d3bd7b6c0c9b4d3c3f8531f30583d1e2/src/netCDF4/_netCDF4.pyx#L1590) in some of their variables. These can be automatically decoded, but only if no `_FillValue` attribute added. So the issue here is not per se that xarray adds `_FillValue=nan`, but that it adds the `_FillValue` attribute in the first place. Not sure how to avoid it, since it is probably intended behaviour by xarray.
